Sources/MongoSwift/MongoCollection+BulkWrite.swift

```diff
@@ -17,118 +17,385 @@ extension MongoCollection {
+public struct BulkWriteResult {
+    /// Number of documents deleted.
+    public let deletedCount: Int
+
     /// Number of documents inserted.
     public let insertedCount: Int
 
     /// Map of the index of the operation to the id of the inserted document.
-    public let insertedIds: [Int: AnyBsonValue]
+    public let insertedIds: [Int: BsonValue?]
 
     /// Number of documents matched for update.
     public let matchedCount: Int
 
     /// Number of documents modified.
     public let modifiedCount: Int
 
-    /// Number of documents deleted.
-    public let deletedCount: Int
-
     /// Number of documents upserted.
     public let upsertedCount: Int
 
     /// Map of the index of the operation to the id of the upserted document.
-    public let upsertedIds: [Int: AnyBsonValue]
+    public let upsertedIds: [Int: BsonValue?]
+
+    fileprivate var writeErrors: [WriteError] = []
+    fileprivate var writeConcernError: WriteConcernError?
+
+    /**
+     * Create a BulkWriteResult operation from a reply and map of inserted IDs.
+     *
+     * Note: we forgo using a Decodable initializer because we still need to
+     * build a map for `upsertedIds` and explicitly add `insertedIds`. While
+     * `mongoc_bulk_operation_execute()` guarantees that `reply` will be
+     * initialized, it doesn't guarantee that all fields will be set. On error,
+     * we should expect fields to be missing and handle that gracefully.
+     *
+     * - Parameters:
+     *   - reply: A `Document` result from `mongoc_bulk_operation_execute()`
+     *   - insertedIds: Map of inserted IDs
+     */
+    fileprivate init(reply: Document, insertedIds: [Int: BsonValue?]) throws {
+        self.deletedCount = reply["nRemoved"] as? Int ?? 0
+        self.insertedCount = reply["nInserted"] as? Int ?? 0
+        self.insertedIds = insertedIds
+        self.matchedCount = reply["nMatched"] as? Int ?? 0
+        self.modifiedCount = reply["nModified"] as? Int ?? 0
+        self.upsertedCount = reply["nUpserted"] as? Int ?? 0
+
+        var upsertedIds = [Int: BsonValue?]()
+
+        if let upserted = reply["upserted"] as? [Document] {
+            for upsert in upserted {
+                guard let index = upsert["index"] as? Int else {
+                    throw MongoError.typeError(message: "Could not cast upserted index to `Int`")
+                }
+                upsertedIds[index] = upsert["_id"]
+            }
+        }
+
+        self.upsertedIds = upsertedIds
+
+        if let writeErrors = reply["writeErrors"] as? [Document] {
+            self.writeErrors = try writeErrors.map { try BsonDecoder().decode(WriteError.self, from: $0) }
+        }
+
+        if let writeConcernErrors = reply["writeConcernErrors"] as? [Document], writeConcernErrors.indices.contains(0) {
+            self.writeConcernError = try BsonDecoder().decode(WriteConcernError.self, from: writeConcernErrors[0])
+        }
+    }
+}
+
+public struct WriteError: Codable {
```

Comment:
can you add a docstring please?

also, just confirming, we don't expect to use either of these errors any place outside of bulk write, right? (if we did, it could make sense for them to live in the same place as `MongoError`.)